GVFS/GVFS.Common/ModifiedPathsDatabase.cs

```diff
@@ -75,38 +75,103 @@ public bool TryAdd(string path, bool isFolder, out bool isRetryable)
-
+                    this.TraceWarning(isFolder, entry, e, nameof(this.TryAdd));
                     return false;
                 }
                 catch (Exception e)
                 {
-                    if (this.Tracer != null)
-                    {
-                        EventMetadata metadata = new EventMetadata();
-                        metadata.Add("Area", "ModifiedPathsDatabase");
-                        metadata.Add(nameof(entry), entry);
-                        metadata.Add(nameof(isFolder), isFolder);
-                        metadata.Add("Exception", e.ToString());
-                        this.Tracer.RelatedError(metadata, $"Exception caught while processing {nameof(this.TryAdd)}");
-                    }
+                    this.TraceError(isFolder, entry, e, nameof(this.TryAdd));
+                    isRetryable = false;
+                    return false;
+                }
 
+                return true;
+            }
+
+            return true;
+        }
+
+        public bool TryRemove(string path, bool isFolder, out bool isRetryable)
+        {
+            isRetryable = true;
+            string entry = this.NormalizeEntryString(path, isFolder);
+            if (this.modifiedPaths.Contains(entry))
+            {
+                isRetryable = true;
+                try
+                {
+                    this.WriteRemoveEntry(entry, () => this.modifiedPaths.TryRemove(entry));
+                }
+                catch (IOException e)
+                {
+                    this.TraceWarning(isFolder, entry, e, nameof(this.TryRemove));
+                    return false;
+                }
+                catch (Exception e)
+                {
+                    this.TraceError(isFolder, entry, e, nameof(this.TryRemove));
                     isRetryable = false;
                     return false;
                 }
+
+                return true;
             }
 
             return true;
         }
 
+        public void WriteAllEntriesAndFlush()
+        {
+            try
+            {
+                this.WriteAndReplaceDataFile(this.GenerateDataLines);
+            }
+            catch (Exception e)
+            {
+                throw new FileBasedCollectionException(e);
+            }
+        }
+
+        private static EventMetadata CreateEventMetadata(bool isFolder, string entry, Exception e)
+        {
+            EventMetadata metadata = new EventMetadata();
+            metadata.Add("Area", "ModifiedPathsDatabase");
+            metadata.Add(nameof(entry), entry);
+            metadata.Add(nameof(isFolder), isFolder);
+            if (e != null)
+            {
+                metadata.Add("Exception", e.ToString());
+            }
+
+            return metadata;
+        }
+
+        private IEnumerable<string> GenerateDataLines()
+        {
+            foreach (string entry in this.modifiedPaths)
+            {
+                yield return this.FormatAddLine(entry);

```

Comment:
What is the perf overhead of the `yield return` syntax if there are many modified paths? Would it be more efficient to switch to `Select` instead? Even better, could we just pass the list of paths to the underlying method and have it write them out in a tight loop? (Assuming the perf hit is meaningful here)

Reply:
`Select` will in effect do the same thing as here.  See [Enumerable.SelectIterator](https://github.com/Microsoft/referencesource/blob/60a4f8b853f60a424e36c7bf60f9b5b5f1973ed1/System.Core/System/Linq/Enumerable.cs#L53).  So I don't see that being different in perf.  The `PlaceholderListDatabase` is doing more work and has been fast enough.  This is essentially a tight loop since the yield return will be the `in` for the loop in `TryWriteTempFile`

```
using (Stream tempFile = this.fileSystem.OpenFileStream(this.tempFilePath, FileMode.Create, FileAccess.Write, FileShare.None, callFlushFileBuffers: true))
using (StreamWriter writer = new StreamWriter(tempFile))
{
	foreach (string line in getDataLines())
	{
		writer.Write(line + NewLine);
	}

	tempFile.Flush();
}
```

I will do some measuring and see if I need to look into this.

Reply:
What `yield return` does is generate a class that implements `IEnumerator` and maintains and updates several fields every time you call `MoveNext`. So it's definitely not free, compared to a `foreach` over an array or list, for example. It could be that this is not a significant hit in this usage, but it would be nice to confirm.

Reply:
I'm not seeing any significant performance difference and couldn't find anything online with performance numbers.  I will add some tracing around this so that we can at least see the numbers if it does cause an issue.